# 1. Initial Analysis

In this notebook we explore the raw data file `data/all_car_adverts.csv`. We make some obvious observations which answer some of the following questions:

- What is this dataset for?
- What columns are in the dataset?

We also state any initial observations which may prove useful in Data Wrangling.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../data/all_car_adverts.csv")
df.head()

,Unnamed: 0,make,model,variant,car_price,car_badges,car_title,car_sub_title,car_attention_grabber,car_specs,...,num_owner,ulez,full_service,part_service,part_warranty,full_dealership,first_year_road_tax,brand_new,finance_available,discounted
0,0,AC,Cobra,NaN,89995.0,NaN,AC Cobra,4.9 MK IV CRS 2dr,GENUINE AC COBRA CRS 522 BHP,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",...,5.0,0,0,0,0,0,0,0,0,0
1,1,AC,Cobra,NaN,92500.0,'',AC Cobra,378 - MkIV,PHYSICAL CAR!,"2019 (T reg), Convertible, 650 miles, Manual, ...",...,NaN,0,0,0,0,0,0,0,0,0
2,2,AC,Cobra,NaN,109995.0,'',AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,FULL CARBON FIBRE BODY.,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",...,3.0,0,0,0,0,0,0,0,0,0
3,3,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV 2dr,ABSOLUTELY STUNNING,"1989 (F reg), Convertible, 2,750 miles, Manual...",...,NaN,0,0,0,0,0,0,0,0,0
4,4,AC,Cobra,NaN,124950.0,'',AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,'STAGE 3' SVO ENGINE,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",...,NaN,0,0,0,0,0,0,0,0,0


In [3]:
df.shape

(818456, 32)

## 1.1. What is this dataset?
According to the [dataset description on Kaggle](https://www.kaggle.com/datasets/guanhaopeng/uk-used-car-market), the dataset is made up of scraped data from a used car website in October 2022. It consists of ~800k records with 32 variables. We note that the exact origin of the dataset is unknown and undocumented. However, for the sake of this project, we are not too concerned with the data being "real".

## 1.2. What columns are in the dataset?



In [4]:
df.columns.values

<ArrowStringArray>
[           'Unnamed: 0',                  'make',                 'model',
               'variant',             'car_price',            'car_badges',
             'car_title',         'car_sub_title', 'car_attention_grabber',
             'car_specs',            'car_seller',     'car_seller_rating',
   'car_seller_location',                  'year',                   'reg',
             'body_type',                 'miles',            'engine_vol',
           'engine_size',      'engine_size_unit',          'transmission',
             'feul_type',             'num_owner',                  'ulez',
          'full_service',          'part_service',         'part_warranty',
       'full_dealership',   'first_year_road_tax',             'brand_new',
     'finance_available',            'discounted']
Length: 32, dtype: str

The columns can be separated into 2 subsets:

1. **Car Listing**: Data that is clearly taken directly from the advertisement (e.g. `car_title`, `car_subtitle`, `car_badges`, `car_attention_grabber`, `car_specs`)
2. **Vehicle**: Data that is about the vehicle itself sans the marketing spiel (e.g. `make`, `model`, `variant`, `engine_size`)

However, looking at the head of the dataframe above we can see that `car_title` is potentially a simple concatenation of the `model` and `variant` and `car_specs` acts as a string representing the columns of subset 2. The question remains as to whether _all_ `car_title` and `car_specs` entries are structured in the same way and if they are not, are there any consequences of this reflected in the dataset?

### 1.2.1. `car_title` Structure

We investigate the structure of `car_title`. We check if it simply a concatenation of some of the other columns, namely `make`, `model` and `variant`.

In [5]:
title_check = df[["make", "model", "variant", "car_title", "car_sub_title", "car_specs"]]
title_check.head(20)

,make,model,variant,car_title,car_sub_title,car_specs
0,AC,Cobra,NaN,AC Cobra,4.9 MK IV CRS 2dr,"2001 (X reg), Convertible, 14,400 miles, 4.9L,..."
1,AC,Cobra,NaN,AC Cobra,378 - MkIV,"2019 (T reg), Convertible, 650 miles, Manual, ..."
2,AC,Cobra,NaN,AC Cobra,MK1V 212 SC. 3.5 V8 350 BHP LOTUS TWIN TURBO. ...,"2000 (X reg), Convertible, 21,600 miles, 3.5L,..."
3,AC,Cobra,NaN,AC Cobra,302 MKIV 2dr,"1989 (F reg), Convertible, 2,750 miles, Manual..."
4,AC,Cobra,NaN,AC Cobra,302 MKIV With Factory Lightweight Engine 5.0 2dr,"1989 (E reg), Convertible, 15,142 miles, 5.0L,..."
5,AC,Cobra,NaN,AC Cobra,378 Superblower 2dr,"2022 (22 reg), Convertible, 500 miles, Manual,..."
6,AC,NaN,MK IV,AC Cobra,302 MKIV 2dr,"1989 (F reg), Convertible, 2,750 miles, Manual..."
7,AC,NaN,MK IV CRS,AC Cobra,4.9 MK IV CRS 2dr,"2001 (X reg), Convertible, 14,400 miles, 4.9L,..."
8,Abarth,124 Spider,GT,Abarth 124 Spider,1.4 T MultiAir 2dr Auto,"2019 (19 reg), Convertible, 10,313 miles, 1.4L..."
9,Abarth,124 Spider,GT,Abarth 124 Spider,1.4 T MultiAir 2dr Auto,"2019 (19 reg), Convertible, 10,313 miles, 1.4L..."


In [6]:
title_check.tail(20)

,make,model,variant,car_title,car_sub_title,car_specs
818436,Volvo,XC90,h T8,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Bright Auto ...,"SUV, 2.0L, 455PS, Automatic, Petrol Plug-in Hy..."
818437,Volvo,XC90,h T8,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Dark Auto 4W...,"2022, SUV, 2.0L, 455PS, Automatic, Petrol Plug..."
818438,Volvo,XC90,h T8,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Bright SUV 5...,"SUV, 2.0L, 455PS, Automatic, Petrol Plug-in Hy..."
818439,Volvo,XC90,h T8,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Dark Auto 4W...,"2022, SUV, 2.0L, 455PS, Automatic, Petrol Plug..."
818440,Volvo,XC90,h T8,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Dark SUV 5dr...,"SUV, 2.0L, 455PS, Automatic, Petrol Plug-in Hy..."
818441,Volvo,XC90,h T8,Volvo XC90,2.0h T8 Recharge 18.8kWh Ultimate Dark Auto 4W...,"SUV, 2.0L, 455PS, Automatic, Petrol Plug-in Hy..."
818442,Westfield,Roadster,NaN,Westfield Roadster,NaN,"2020 (20 reg), 3,000 miles, Manual, Full servi..."
818443,Westfield,SE,NaN,Westfield SE,2.0 2dr,"1999 (T reg), Convertible, 16,362 miles, 2.0L,..."
818444,Westfield,SE,NaN,Westfield SE,2.0 2dr,"1991 (H reg), Convertible, 40,000 miles, 2.0L,..."
818445,Westfield,SEight,NaN,Westfield Seight,3.9 Sports 2dr,"1998 (R reg), Convertible, 15,500 miles, 3.9L,..."


Clearly `car_title` does not provide provide the `variant` reliably, only the `make` and `model`. In the cases where it seems the `variant` column is provided, it seems to come from the `car_sub_title`. That being said, I'm not particular convinced the `variant` column is reliable. Columns 11-19 list "MultiAir" as the `variant`. However, this is factually incorrect. "MultiAir" refers to the engine technology; the [Abarth website](https://www.abarth.com/124-spider) refers to the "engine type" as "1.4 Turbo MultiAir 170 HP".

We check this `car_title ~ make + " " + model` hypothesis. We do not assume this relationship to be case sensitive.

In [7]:
to_check = df[~df["make"].isna() & ~df["model"].isna()][["make", "model", "variant", "car_title"]]

# Convert car_title to lower case (avoid case sensitivity)
to_check["car_title"] = to_check["car_title"].str.lower()

# Construct "assumed" title
to_check["assumed"] = (to_check["make"] + " " + to_check["model"]).str.lower()

to_check.head(20)

,make,model,variant,car_title,assumed
0,AC,Cobra,NaN,ac cobra,ac cobra
1,AC,Cobra,NaN,ac cobra,ac cobra
2,AC,Cobra,NaN,ac cobra,ac cobra
3,AC,Cobra,NaN,ac cobra,ac cobra
4,AC,Cobra,NaN,ac cobra,ac cobra
5,AC,Cobra,NaN,ac cobra,ac cobra
8,Abarth,124 Spider,GT,abarth 124 spider,abarth 124 spider
9,Abarth,124 Spider,GT,abarth 124 spider,abarth 124 spider
10,Abarth,124 Spider,GT,abarth 124 spider,abarth 124 spider
11,Abarth,124 Spider,MultiAir,abarth 124 spider,abarth 124 spider


In [8]:
mask = to_check["car_title"] != to_check["assumed"]
no_match = to_check[mask]

print(f"#Mismatches = {len(no_match)}")
print(f"% Mismatches = {mask.mean()}")

no_match.drop_duplicates()

#Mismatches = 87556
% Mismatches = 0.10701665825340737


,make,model,variant,car_title,assumed
12,Abarth,124 Spider,MultiAir,abarth 124,abarth 124 spider
70,Abarth,124 Spider,Scorpione,abarth 124,abarth 124 spider
160,Abarth,500C,T-Jet,abarth 500,abarth 500c
181,Abarth,595,70th,abarth 500,abarth 595
288,Abarth,595,Competizione,abarth 500,abarth 595
...,...,...,...,...,...
818200,Volvo,XC90,h T8,volvo xc90 t8 twin engine awd inscription pro,volvo xc90
818234,Volvo,XC90,h T8,"volvo xc90 ii recharge inscription, t8 awd",volvo xc90
818255,Volvo,XC90,h T8,"volvo xc90 ii recharge r-design, t8 awd",volvo xc90
818379,Volvo,XC90,h T8,"volvo xc90 recharge inscription pro, t8 awd",volvo xc90


Clearly, there are some instances in which our hypothesis does not hold. These account for ~10.7% of all records in the dataset. Some of these are duplicate cases. This means that we must be careful if we attempt to recover `make` or `model` data using the `car_title` column.

### 1.2.2 `car_specs` Structure

We investigate the structure of strings in the `car_specs` column compared with some of the other columns which provide details about the car.

In [9]:
car_check = df[["car_specs", "year", "reg", "body_type", 'miles', 'engine_vol', 'engine_size', 'engine_size_unit', 'transmission', 'feul_type', 'num_owner', 'ulez']]
car_check.head(20)

,car_specs,year,reg,body_type,miles,engine_vol,engine_size,engine_size_unit,transmission,feul_type,num_owner,ulez
0,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",2001,X reg,convertible,14400.0,4.9,225.0,bhp,manual,petrol,5.0,0
1,"2019 (T reg), Convertible, 650 miles, Manual, ...",2019,T reg,convertible,650.0,NaN,NaN,NaN,manual,petrol,NaN,0
2,"2000 (X reg), Convertible, 21,600 miles, 3.5L,...",2000,X reg,convertible,21600.0,3.5,NaN,NaN,manual,petrol,3.0,0
3,"1989 (F reg), Convertible, 2,750 miles, Manual...",1989,F reg,convertible,2750.0,NaN,NaN,NaN,manual,petrol,NaN,0
4,"1989 (E reg), Convertible, 15,142 miles, 5.0L,...",1989,E reg,convertible,15142.0,5.0,NaN,NaN,manual,petrol,NaN,0
5,"2022 (22 reg), Convertible, 500 miles, Manual,...",2022,22 reg,convertible,500.0,NaN,NaN,NaN,manual,petrol,NaN,0
6,"1989 (F reg), Convertible, 2,750 miles, Manual...",1989,F reg,convertible,2750.0,NaN,NaN,NaN,manual,petrol,NaN,0
7,"2001 (X reg), Convertible, 14,400 miles, 4.9L,...",2001,X reg,convertible,14400.0,4.9,225.0,bhp,manual,petrol,5.0,0
8,"2019 (19 reg), Convertible, 10,313 miles, 1.4L...",2019,19 reg,convertible,10313.0,1.4,167.0,bhp,automatic,petrol,2.0,1
9,"2019 (19 reg), Convertible, 10,313 miles, 1.4L...",2019,19 reg,convertible,10313.0,1.4,167.0,bhp,automatic,petrol,2.0,1


In [10]:
car_check.tail(20)

,car_specs,year,reg,body_type,miles,engine_vol,engine_size,engine_size_unit,transmission,feul_type,num_owner,ulez
818436,"SUV, 2.0L, 455PS, Automatic, Petrol Plug-in Hy...",SUV,NaN,suv,NaN,2.0,455.0,ps,automatic,petrol plug-in hybrid,NaN,1
818437,"2022, SUV, 2.0L, 455PS, Automatic, Petrol Plug...",2022,NaN,suv,NaN,2.0,455.0,ps,automatic,petrol plug-in hybrid,NaN,1
818438,"SUV, 2.0L, 455PS, Automatic, Petrol Plug-in Hy...",SUV,NaN,suv,NaN,2.0,455.0,ps,automatic,petrol plug-in hybrid,NaN,1
818439,"2022, SUV, 2.0L, 455PS, Automatic, Petrol Plug...",2022,NaN,suv,NaN,2.0,455.0,ps,automatic,petrol plug-in hybrid,NaN,1
818440,"SUV, 2.0L, 455PS, Automatic, Petrol Plug-in Hy...",SUV,NaN,suv,NaN,2.0,455.0,ps,automatic,petrol plug-in hybrid,NaN,1
818441,"SUV, 2.0L, 455PS, Automatic, Petrol Plug-in Hy...",SUV,NaN,suv,NaN,2.0,455.0,ps,automatic,petrol plug-in hybrid,NaN,1
818442,"2020 (20 reg), 3,000 miles, Manual, Full servi...",2020,20 reg,NaN,3000.0,NaN,NaN,NaN,manual,NaN,NaN,0
818443,"1999 (T reg), Convertible, 16,362 miles, 2.0L,...",1999,T reg,convertible,16362.0,2.0,NaN,NaN,manual,petrol,2.0,0
818444,"1991 (H reg), Convertible, 40,000 miles, 2.0L,...",1991,H reg,convertible,40000.0,2.0,NaN,NaN,manual,petrol,NaN,0
818445,"1998 (R reg), Convertible, 15,500 miles, 3.9L,...",1998,R reg,convertible,15500.0,3.9,300.0,bhp,manual,petrol,NaN,0


Some of the columns consist of errors, e.g. `year` on row 818436 is "SUV". Therefore, we theorise that the other columns in the dataset were scraped from `car_specs` using a set of strict rules. We look at rows 7 and 15, which contains a value in every column, to reverse engineer these rules.

In [11]:
car_check.loc[7, "car_specs"]

'2001 (X reg), Convertible, 14,400 miles, 4.9L, 225BHP, Manual, Petrol, 5 owners'

In [12]:
car_check.loc[15, "car_specs"]

'2017 (67 reg), Convertible, 37,098 miles, 1.4L, 167BHP, Manual, Petrol, 2 owners, ULEZ'

Clearly the `car_specs` column is assumed to be some string of comma separated values which inform the inputs of the other columns in the dataframe. By performing iterative testing using regex, we found that all rows, without any missing data, have a `car_specs` value which follows a pattern:

`YEAR (REG), BODY_TYPE, MILEAGE, ENGINE_VOLUME, POWER, TRANSMISSION, FUEL, OWNERS, (ADDITIONAL STRINGS ABOUT SERVICES/SERVICE HISTORY), ULEZ?`

We describe this in more detail in a regex below.

In [13]:
pattern = (
    r"^\d{4} \([A-Z0-9]+ reg\), "            # Year (reg)
    r".+, "                                  # Body type
    r"[\d,]+ miles, "                        # Mileage
    r"\d+(\.\d+)?L, "                        # Engine volume
    r"\d+(\.\d+)?(BHP|PS|bhp|ps), "          # Power
    r".+, "                                  # Transmission
    r".+, "                                  # Fuel
    r"\d+ owner(?:s)?"                       # Owner(s)
    r"(, No (S|s)ervice (H|h)istory)?"       # Full Service History
    r"(, Full (S|s)ervice (H|h)istory)?"     # Full Service History
    r"(, Part (S|s)ervice (H|h)istory)?"     # Part Service History
    r"(, Full (D|d)ealership (H|h)istory)?"  # Full Dealership History
    r"(, First service is not due)?"         # First service isn't due
    r"(, ULEZ)?$"                            # Optional ULEZ
)

matches = car_check["car_specs"].str.match(pattern, na=False)
print(f"#Regex Matches = {matches.sum()}")
print(f"#Regex Mismatches = {len(matches) - matches.sum()}")

#Regex Matches = 424797
#Regex Mismatches = 393659


In [14]:
contains_na = car_check[
    car_check.isna().any(axis=1)
]

print(f"#Rows containing NaN: {len(contains_na)}")
print(f"#Matches containing NaN: {len(contains_na[matches])}")
print(f"#Mismatches containing NaN: {len(contains_na[~matches])}")

#Rows containing NaN: 393659
#Matches containing NaN: 0
#Mismatches containing NaN: 393659


/tmp/ipykernel_367329/2277219315.py:6: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  print(f"#Matches containing NaN: {len(contains_na[matches])}")
/tmp/ipykernel_367329/2277219315.py:7: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  print(f"#Mismatches containing NaN: {len(contains_na[~matches])}")


This in conclusion, there are precisely 393,659 rows which contain a N/A value in one of the columns we have been inspecting, and **all** of them do not match the regex of our description of `car_stats`.

### 1.2.3. `car_badges` Structure

In this section, we repeat a similar analysis of `car_badges` as we did for `car_specs` and `car_title` to see how the content of this column is reflected in the other columns in the dataset.

In [15]:
df["car_badges"].unique()

<ArrowStringArray>
[                                                           nan,
                                                           '''',
                                          ''Finance available'',
           ''Good price', 'Approved used', 'Finance available'',
          ''Great price', 'Approved used', 'Finance available'',
                           ''Lower price', 'Finance available'',
                            ''Good price', 'Finance available'',
 ''Good price', 'Finance available', 'Part-ex price guarantee'',
                                                 ''Good price'',
                           ''Great price', 'Finance available'',
 ...
                                  ''Save £10,616', 'Brand new'',
              ''Save £4,855', 'Brand new', 'Finance available'',
                                   ''Save £7,866', 'Brand new'',
              ''Save £7,855', 'Brand new', 'Finance available'',
              ''Save £2,592', 'Brand new', 'Finance available'',
 

It seems that `car_badges` contains a string of comma separated strings. Some of these strings are directly reflected in the dataset, e.g. `discounted`, `finance_available`.

In [16]:
unique_badges = (
    df["car_badges"]
        .dropna()
        .str.split("',")
        .explode()
        .str.replace("'", "")
        .str.strip()
        .unique()
)

unique_badges

<ArrowStringArray>
[                       '',       'Finance available',
              'Good price',           'Approved used',
             'Great price',             'Lower price',
 'Part-ex price guarantee',              'Fair price',
                   'Cat S',                   'Cat D',
 ...
             'Save £7,269',            'Save £10,356',
            'Save £10,631',            'Save £10,616',
             'Save £7,866',             'Save £7,855',
             'Save £8,191',            'Save £10,091',
            'Save £11,158',             'Save £4,926']
Length: 3425, dtype: str

A lot of these badges have form "Save £x", which is not particularly useful for what we are looking for. Similar things can be said about the badges of form "Cat x".

In [17]:
unique_badges = [
    x for x in unique_badges
    if not x.lower().startswith(("save", "cat"))
]

unique_badges

['',
 'Finance available',
 'Good price',
 'Approved used',
 'Great price',
 'Lower price',
 'Part-ex price guarantee',
 'Fair price',
 'Higher price',
 'Brand new']

Thus it is clear that entries in this column inform the values of `brand_new`, `discounted`, `finance_available`.

### 1.2.4. Column Definitions

We have sufficiently studied the dataset in order to define what each of the columns mean in the context of the dataset.

- `Unnamed: 0` - an index column of some kind
- `make` - the make of the car
- `model` - the car model
- `variant` - the variant of model
- `car_price` - the price of the car
- `car_badges` - a list of comma separated strings to help with car advertisement (e.g. 'good price', 'finance available')
- `car_title` - the car's title, usually a concatenation of `make` and `model`
- `car_sub_title` - additional details of the car, usually detailing the engine and model variant
- `car_attention_grabber` - a tagline for the car sale listing
- `car_specs` - a comma separated list of car specifications, data from this column can be found in other columns, where available
- `car_seller` - company which is selling the car
- `car_seller_rating` - rating of the company selling the car
- `car_seller_location` - where the company selling the car is based
- `year` - the year the car was registered
- `reg` - the character(s) related to the cars registration*
- `body_type` - the body type of the car
- `miles` - car milage
- `engine_vol` - the volume of the car engine in litres
- `engine_size` - the power of the car (arbitrary units)
- `engine_size_unit` - the units used for `engine_size`
- `transmission` - the type of transmission e.g. manual, automatic
- `feul_type` - type of fuel used for the car, e.g. petrol, diesel
- `num_owner` - the number of owners the car has had
- `ulez` - whether or not the car meets the standards to drive in London's "Ultra Low Emission Zone" without being charged
- `full_service` - whether or not the car has a full service history
- `part_service` - whether or not the car has a partial service history
- `part_warranty` - whether or not some of the parts of the car are covered by some form of warranty
- `full_dealership` - whether or not the car has a full dealership history
- `first_year_road_tax` - whether or not the first year of road tax is offered as part of the sales deal
- `brand_new` - whether or not the car is brand new
- `finance_available` - whether or not the car is purchasable with finance
- `discounted` - whether or not the car is offered at a discounted price

*in the UK, there is typically 2 character(s) relating to each year, and this details the semester in which the car was registered.